File: sdks/typescript/src/mob.ts
import { EventSubscription } from "./subscription.js";
import type {
  AgentEventEnvelope,
  AttributedMobEvent,
  ContentBlock,
  MobEventsOptions,
  MobEventsResult,
  MobFlowStatus,
  MobLifecycleAction,
  MobMember,
  MobMemberRef,
  MobRunResult,
  MobWireMembersBatchEdgeInput,
  MobWireMembersBatchResult,
  ResolvedModelCapabilities,
  MobSpawnResult,
  MobStatus,
  SpawnManySpec,
  SpawnSpec,
} from "./types.js";
import type {
  MobSpawnManyResultEntry,
  WireAuthBindingRef,
  WireMobMemberStatus,
  WireMemberProgressSnapshot,
} from "./generated/types.js";
import type { MeerkatClient } from "./client.js";

export type MobHandlingMode = "queue" | "steer";
export type MobRenderClass =
  | "user_prompt"
  | "peer_message"
  | "peer_request"
  | "peer_response"
  | "external_event"
  | "flow_step"
  | "continuation"
  | "system_notice"
  | "tool_scope_notice"
  | "ops_progress";

export interface MobRenderMetadata extends Record<string, unknown> {
  class: MobRenderClass;
  salience?: "background" | "normal" | "important" | "urgent";
}

export interface MemberSendOptions {
  handlingMode?: MobHandlingMode;
  renderMetadata?: MobRenderMetadata;
}

export interface MemberDeliveryReceipt {
  mobId: string;
  agentIdentity: string;
  memberRef: MobMemberRef;
  handlingMode: MobHandlingMode;
}

export interface MemberRespawnReceipt {
  agentIdentity: string;
  memberRef: MobMemberRef;
}

export interface MobRespawnResult {
  status: "completed" | "topology_restore_failed";
  receipt: MemberRespawnReceipt;
  failedPeerIds?: string[];
}

export interface MobUnreachablePeer {
  peer: string;
  reason?: string;
}

export interface MobPeerConnectivitySnapshot {
  reachablePeerCount: number;
  unknownPeerCount: number;
  unreachablePeers: MobUnreachablePeer[];
}

/**
 * Tri-state peer-connectivity projection for a mob member snapshot.
 *
 * Distinguishes a member with no bridge session (`not_applicable`) from a
 * transiently-unresolved live probe (`probe_timed_out`) from a resolved
 * connectivity snapshot (`known`).
 */
export type MobPeerConnectivity =
  | { status: "not_applicable" }
  | { status: "probe_timed_out" }
  | { status: "known"; snapshot: MobPeerConnectivitySnapshot };

export interface MobMemberSnapshot {
  status: WireMobMemberStatus;
  /** Server-resolved opaque handle for subsequent member-targeted calls. */
  memberRef: MobMemberRef;
  outputPreview?: string;
  error?: string;
  tokensUsed: number;
  isFinal: boolean;
  currentSessionId?: string;
  resolvedCapabilities?: ResolvedModelCapabilities;
  kickoff?: Record<string, unknown>;
  externalMember?: unknown;
  peerConnectivity?: MobPeerConnectivity;
  progress?: WireMemberProgressSnapshot;
}

export interface MobKickoffWaitOptions {
  memberIds?: string[];
  timeoutMs?: number;
}

/**
 * Member snapshot carried by the `mob/wait_kickoff` / `mob/wait_ready`
 * member lists. The wait wire carries no `member_ref`, so this is its own
 * shape rather than an extension of {@link MobMemberSnapshot}.
 */
export interface MobKickoffMemberSnapshot {
  agentIdentity: string;
  status: string;
  outputPreview?: string;
  error?: string;
  tokensUsed: number;
  isFinal: boolean;
  peerConnectivity?: MobPeerConnectivity;
}

export type MobReadyWaitOptions = MobKickoffWaitOptions;
export type MobReadyMemberSnapshot = MobKickoffMemberSnapshot;

export interface ExternalPeerTarget {
  readonly external: {
    readonly name: string;
    readonly address: string;
    readonly identity: {
      readonly kind: "ed25519_public_key";
      readonly public_key: string;
    };
  };
}

export type MobPeerTarget = string | ExternalPeerTarget;

export interface MobHelperResult {
  output?: string;
  tokensUsed: number;
  agentIdentity: string;
  memberRef: MobMemberRef;
}

export class Member {
  private readonly client: MeerkatClient;
  readonly mobId: string;
  readonly agentIdentity: string;

  /** @internal */
  constructor(client: MeerkatClient, mobId: string, agentIdentity: string) {
    this.client = client;
    this.mobId = mobId;
    this.agentIdentity = agentIdentity;
  }

  async send(
    content: string | ContentBlock[],
    options?: MemberSendOptions,
  ): Promise<MemberDeliveryReceipt> {
    return this.client.sendMobMemberContent(this.mobId, this.agentIdentity, content, options);
  }

  async events(): Promise<EventSubscription<AgentEventEnvelope>> {
    return this.client.subscribeMobMemberEvents(this.mobId, this.agentIdentity);
  }
}

export class Mob {
  private readonly client: MeerkatClient;
  readonly mobId: string;

  /** @internal */
  constructor(client: MeerkatClient, mobId: string) {
    this.client = client;
    this.mobId = mobId;
  }

  async status(): Promise<MobStatus> {
    return this.client.mobStatus(this.mobId);
  }

  async lifecycle(action: MobLifecycleAction): Promise<void> {
    await this.client.mobLifecycle(this.mobId, action);
  }

  async spawn(spec: SpawnSpec): Promise<MobSpawnResult> {
    return this.client.spawnMobMember(this.mobId, spec);
  }

  async spawnMany(specs: SpawnManySpec[]): Promise<MobSpawnManyResultEntry[]> {
    return this.client.spawnMobMembers(this.mobId, specs);
  }

  async retire(agentIdentity: string): Promise<void> {
    await this.client.retireMobMember(this.mobId, agentIdentity);
  }

  async respawn(
    agentIdentity: string,
    initialMessage?: string | ContentBlock[],
  ): Promise<MobRespawnResult> {
    return this.client.respawnMobMember(this.mobId, agentIdentity, initialMessage);
  }

  async forceCancel(agentIdentity: string): Promise<void> {
    await this.client.forceCancelMobMember(this.mobId, agentIdentity);
  }

  async memberStatus(agentIdentity: string): Promise<MobMemberSnapshot> {
    return this.client.mobMemberStatus(this.mobId, agentIdentity);
  }

  async waitForKickoffComplete(
    options?: MobKickoffWaitOptions,
  ): Promise<MobKickoffMemberSnapshot[]> {
    return this.client.waitMobKickoff(this.mobId, options);
  }

  async waitForReady(
    options?: MobReadyWaitOptions,
  ): Promise<MobReadyMemberSnapshot[]> {
    return this.client.waitMobReady(this.mobId, options);
  }

  async spawnHelper(
    prompt: string,
    options?: {
      agentIdentity?: string;
      roleName?: string;
      profileName?: string;
      modelOverride?: string;
      authBinding?: WireAuthBindingRef;
      runtimeMode?: string;
      backend?: string;
    },
  ): Promise<MobHelperResult> {
    return this.client.spawnMobHelper(this.mobId, prompt, options);
  }

  async forkHelper(
    sourceMemberId: string,
    prompt: string,
    options?: {
      agentIdentity?: string;
      roleName?: string;
      profileName?: string;
      modelOverride?: string;
      authBinding?: WireAuthBindingRef;
      forkContext?: Record<string, unknown>;
      runtimeMode?: string;
      backend?: string;
    },
  ): Promise<MobHelperResult> {
    return this.client.forkMobHelper(this.mobId, sourceMemberId, prompt, options);
  }

  async wire(member: string, peer: MobPeerTarget): Promise<void> {
    await this.client.wireMobMembers(this.mobId, member, peer);
  }

  async wireMembersBatch(
    edges: readonly MobWireMembersBatchEdgeInput[],
  ): Promise<MobWireMembersBatchResult> {
    return this.client.mobWireMembersBatch(this.mobId, edges);
  }

  async unwire(member: string, peer: MobPeerTarget): Promise<void> {
    await this.client.unwireMobMembers(this.mobId, member, peer);
  }

  async listMembers(): Promise<MobMember[]> {
    return this.client.listMobMembers(this.mobId);
  }

  member(agentIdentity: string): Member {
    return new Member(this.client, this.mobId, agentIdentity);
  }

  async appendSystemContext(
    agentIdentity: string,
    text: string,
    options?: { source?: string; idempotencyKey?: string },
  ): Promise<Record<string, unknown>> {
    return this.client.appendMobSystemContext(this.mobId, agentIdentity, text, options);
  }

  async listFlows(): Promise<string[]> {
    return this.client.listMobFlows(this.mobId);
  }

  async runFlow(flowId: string, params: Record<string, unknown> = {}): Promise<string> {
    return this.client.runMobFlow(this.mobId, flowId, params);
  }

  async run(
    params: Record<string, unknown> = {},
    options: { prompt?: string; flowId?: string } = {},
  ): Promise<string> {
    return this.client.runMob(this.mobId, params, options);
  }

  async flowStatus(runId: string): Promise<MobFlowStatus | null> {
    return this.client.getMobFlowStatus(this.mobId, runId);
  }

  async runResult(runId: string): Promise<MobRunResult | null> {
    return this.client.getMobRunResult(this.mobId, runId);
  }

  async cancelFlow(runId: string): Promise<void> {
    await this.client.cancelMobFlow(this.mobId, runId);
  }

  async subscribeMemberEvents(agentIdentity: string): Promise<EventSubscription<AgentEventEnvelope>> {
    return this.client.subscribeMobMemberEvents(this.mobId, agentIdentity);
  }

  async subscribeEvents(): Promise<EventSubscription<AttributedMobEvent>> {
    return this.client.subscribeMobEvents(this.mobId);
  }

  async readEvents(options?: MobEventsOptions): Promise<MobEventsResult> {
    return this.client.readMobEvents(this.mobId, options);
  }
}

File: sdks/python/meerkat/mob.py
from __future__ import annotations

from typing import Any, Literal, NotRequired, TypedDict

from .generated.types import (
    MobSpawnManyResult,
    WireBudgetSplitPolicy,
    WireAuthBindingRef,
    WireContentInput,
    WireMemberLaunchMode,
    WireMobBackendKind,
    WireMobProfile,
    WireMobRuntimeMode,
    WireRuntimeBinding,
    WireToolAccessPolicy,
    WireToolFilter,
)
from .streaming import EventSubscription
from .types import ResolvedModelCapabilities

MobLifecycleAction = Literal["stop", "resume", "complete", "reset", "destroy"]

WorkOrigin = Literal["external", "internal"]

# Opaque server-resolved handle for a mob member. App code treats this as
# a transparent token — obtained from `ensure_member`/`spawn_helper`/
# `fork_helper`/member-list responses, passed back on work-lane and
# member-targeted calls.
MobMemberRef = str

RenderClass = Literal[
    "user_prompt",
    "peer_message",
    "peer_request",
    "peer_response",
    "external_event",
    "flow_step",
    "continuation",
    "system_notice",
    "tool_scope_notice",
    "ops_progress",
]
RenderSalience = Literal["background", "normal", "important", "urgent"]
RenderMetadata = TypedDict(
    "RenderMetadata",
    {"class": RenderClass, "salience": NotRequired[RenderSalience]},
)

MemberDeliveryReceipt = TypedDict(
    "MemberDeliveryReceipt",
    {
        "agent_identity": str,
        "member_ref": MobMemberRef,
        "handling_mode": Literal["queue", "steer"],
    },
)

MemberRespawnReceipt = TypedDict(
    "MemberRespawnReceipt",
    {
        "agent_identity": str,
        "member_ref": MobMemberRef,
    },
)

MobSpawnResult = TypedDict(
    "MobSpawnResult",
    {
        "mob_id": str,
        "agent_identity": str,
        "member_ref": MobMemberRef,
    },
)

MobRespawnResult = TypedDict(
    "MobRespawnResult",
    {
        "status": Literal["completed", "topology_restore_failed"],
        "receipt": MemberRespawnReceipt,
        "failed_peer_ids": NotRequired[list[str]],
    },
)

MobUnreachablePeer = TypedDict(
    "MobUnreachablePeer",
    {
        "peer": str,
        "reason": NotRequired[str],
    },
)

MobPeerConnectivitySnapshot = TypedDict(
    "MobPeerConnectivitySnapshot",
    {
        "reachable_peer_count": int,
        "unknown_peer_count": int,
        "unreachable_peers": list[MobUnreachablePeer],
    },
)

MobMemberSnapshot = TypedDict(
    "MobMemberSnapshot",
    {
        "status": str,
        "output_preview": NotRequired[str],
        "error": NotRequired[str],
        "tokens_used": int,
        "is_final": bool,
        "realtime_attachment_status": NotRequired[
            Literal[
                "unattached",
                "intent_present_unbound",
                "binding_not_ready",
                "binding_ready",
                "replacement_pending",
                "reattach_required",
            ]
        ],
        "resolved_capabilities": NotRequired[ResolvedModelCapabilities],
        "peer_connectivity": NotRequired[MobPeerConnectivitySnapshot],
        "kickoff": NotRequired[dict[str, Any]],
        # Diagnostic bridge-session id for status/continuity only. Realtime
        # callers open `RealtimeChannel.mob_member(...)`; the server resolves
        # the current binding behind the stable mob-member target.
        "current_session_id": NotRequired[str],
    },
)

MobKickoffMemberSnapshot = TypedDict(
    "MobKickoffMemberSnapshot",
    {
        "agent_identity": str,
        "status": str,
        "output_preview": NotRequired[str],
        "error": NotRequired[str],
        "tokens_used": int,
        "is_final": bool,
        "peer_connectivity": NotRequired[MobPeerConnectivitySnapshot],
        "kickoff": NotRequired[dict[str, Any]],
    },
)

MobReadyMemberSnapshot = MobKickoffMemberSnapshot

MobHelperResult = TypedDict(
    "MobHelperResult",
    {
        "output": NotRequired[str],
        "tokens_used": int,
        "agent_identity": str,
        "member_ref": MobMemberRef,
    },
)

MobMember = TypedDict(
    "MobMember",
    {
        "agent_identity": str,
        "member_ref": MobMemberRef,
        "profile": str,
        "peer_id": NotRequired[str],
        "external_peer_specs": NotRequired[dict[str, dict[str, Any]]],
        "runtime_mode": NotRequired[str],
        "state": NotRequired[str],
        "wired_to": NotRequired[list[str]],
        "labels": NotRequired[dict[str, str]],
        "status": NotRequired[str],
        "error": NotRequired[str],
        "is_final": NotRequired[bool],
        "kickoff": NotRequired[dict[str, Any]],
    },
)

MobSpawnSpec = TypedDict(
    "MobSpawnSpec",
    {
        "profile": str,
        "agent_identity": str,
        "initial_message": NotRequired[WireContentInput | None],
        "runtime_mode": NotRequired[WireMobRuntimeMode | None],
        "backend": NotRequired[WireMobBackendKind | None],
        "labels": NotRequired[dict[str, str] | None],
        "context": NotRequired[dict[str, Any] | None],
        "additional_instructions": NotRequired[list[str] | None],
        "auth_binding": NotRequired[WireAuthBindingRef | dict[str, str] | None],
    },
)


class Member:
    """Capability-bearing mob member handle."""

    def __init__(self, mob: "Mob", agent_identity: str):
        self._mob = mob
        self.agent_identity = agent_identity

    async def send(
        self,
        content: str | list[dict[str, Any]],
        *,
        handling_mode: Literal["queue", "steer"] = "queue",
        render_metadata: RenderMetadata | None = None,
    ) -> MemberDeliveryReceipt:
        return await self._mob._client.send_mob_member_content(
            self._mob.id,
            self.agent_identity,
            content,
            handling_mode=handling_mode,
            render_metadata=render_metadata,
        )

    async def events(self) -> EventSubscription:
        return await self._mob.subscribe_member_events(self.agent_identity)


class Mob:
    """First-class mob handle backed by explicit RPC methods."""

    def __init__(self, client: Any, mob_id: str):
        self._client = client
        self.id = mob_id

    async def status(self) -> dict[str, Any]:
        return await self._client.mob_status(self.id)

    async def members(self) -> list[MobMember]:
        return await self._client.list_mob_members(self.id)

    async def spawn(
        self,
        *,
        profile: str,
        agent_identity: str,
        initial_message: WireContentInput | None = None,
        runtime_mode: WireMobRuntimeMode | None = None,
        backend: WireMobBackendKind | None = None,
        labels: dict[str, str] | None = None,
        context: dict[str, Any] | None = None,
        additional_instructions: list[str] | None = None,
        binding: WireRuntimeBinding | dict[str, Any] | None = None,
        shell_env: dict[str, str] | None = None,
        auto_wire_parent: bool | None = None,
        launch_mode: WireMemberLaunchMode | dict[str, Any] | None = None,
        tool_access_policy: WireToolAccessPolicy | dict[str, Any] | None = None,
        budget_split_policy: WireBudgetSplitPolicy | dict[str, Any] | None = None,
        inherited_tool_filter: WireToolFilter | dict[str, list[str]] | None = None,
        override_profile: WireMobProfile | dict[str, Any] | None = None,
        auth_binding: WireAuthBindingRef | dict[str, str] | None = None,
    ) -> MobSpawnResult:
        return await self._client.spawn_mob_member(
            self.id,
            profile=profile,
            agent_identity=agent_identity,
            initial_message=initial_message,
            runtime_mode=runtime_mode,
            backend=backend,
            labels=labels,
            context=context,
            additional_instructions=additional_instructions,
            binding=binding,
            shell_env=shell_env,
            auto_wire_parent=auto_wire_parent,
            launch_mode=launch_mode,
            tool_access_policy=tool_access_policy,
            budget_split_policy=budget_split_policy,
            inherited_tool_filter=inherited_tool_filter,
            override_profile=override_profile,
            auth_binding=auth_binding,
        )

    async def spawn_many(
        self,
        specs: list[MobSpawnSpec],
    ) -> MobSpawnManyResult:
        return await self._client.spawn_mob_members(self.id, specs)

    async def read_events(
        self,
        *,
        after_cursor: int = 0,
        limit: int = 100,
    ) -> dict[str, list[dict[str, Any]]]:
        return await self._client.read_mob_events(
            self.id,
            after_cursor=after_cursor,
            limit=limit,
        )

    async def retire(self, agent_identity: str) -> None:
        await self._client.retire_mob_member(self.id, agent_identity)

    async def respawn(
        self,
        agent_identity: str,
        initial_message: str | list[dict[str, Any]] | None = None,
    ) -> MobRespawnResult:
        return await self._client.respawn_mob_member(self.id, agent_identity, initial_message)

    async def force_cancel(self, agent_identity: str) -> None:
        await self._client.force_cancel_mob_member(self.id, agent_identity)

    async def member_status(self, agent_identity: str) -> MobMemberSnapshot:
        return await self._client.mob_member_status(self.id, agent_identity)

    async def wait_for_kickoff_complete(
        self,
        *,
        member_ids: list[str] | None = None,
        timeout_ms: int | None = None,
    ) -> list[MobKickoffMemberSnapshot]:
        return await self._client.wait_mob_kickoff(
            self.id,
            member_ids=member_ids,
            timeout_ms=timeout_ms,
        )

    async def wait_for_ready(
        self,
        *,
        member_ids: list[str] | None = None,
        timeout_ms: int | None = None,
    ) -> list[MobReadyMemberSnapshot]:
        return await self._client.wait_mob_ready(
            self.id,
            member_ids=member_ids,
            timeout_ms=timeout_ms,
        )

    async def spawn_helper(
        self,
        prompt: str,
        *,
        agent_identity: str | None = None,
        role_name: str | None = None,
    ) -> MobHelperResult:
        return await self._client.spawn_mob_helper(
            self.id,
            prompt,
            agent_identity=agent_identity,
            role_name=role_name,
        )

    async def fork_helper(
        self,
        source_member_id: str,
        prompt: str,
        *,
        agent_identity: str | None = None,
        role_name: str | None = None,
        fork_context: dict[str, Any] | None = None,
    ) -> MobHelperResult:
        return await self._client.fork_mob_helper(
            self.id,
            source_member_id,
            prompt,
            agent_identity=agent_identity,
            role_name=role_name,
            fork_context=fork_context,
        )

    async def wire(self, member: str, peer: str | dict[str, Any]) -> None:
        await self._client.wire_mob_members(self.id, member, peer)

    async def unwire(self, member: str, peer: str | dict[str, Any]) -> None:
        await self._client.unwire_mob_members(self.id, member, peer)

    async def lifecycle(self, action: str) -> None:
        await self._client.mob_lifecycle(self.id, action)

    def member(self, agent_identity: str) -> Member:
        return Member(self, agent_identity)

    async def append_system_context(
        self,
        agent_identity: str,
        text: str,
        *,
        source: str | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        return await self._client.append_mob_system_context(
            self.id,
            agent_identity,
            text,
            source=source,
            idempotency_key=idempotency_key,
        )

    async def flows(self) -> list[str]:
        return await self._client.list_mob_flows(self.id)

    async def run_flow(self, flow_id: str, params: dict[str, Any] | None = None) -> str:
        return await self._client.run_mob_flow(self.id, flow_id, params or {})

    async def flow_status(self, run_id: str) -> dict[str, Any] | None:
        return await self._client.get_mob_flow_status(self.id, run_id)

    async def cancel_flow(self, run_id: str) -> None:
        await self._client.cancel_mob_flow(self.id, run_id)

    async def subscribe_member_events(self, agent_identity: str) -> EventSubscription:
        """Subscribe to events for a single mob member."""
        return await self._client.subscribe_mob_member_events(self.id, agent_identity)

    async def subscribe_events(self) -> EventSubscription:
        """Subscribe to attributed mob-wide events."""
        return await self._client.subscribe_mob_events(self.id)

    async def subscribe_member(self, agent_identity: str) -> EventSubscription:
        return await self.subscribe_member_events(agent_identity)

    async def subscribe_all(self) -> EventSubscription:
        return await self.subscribe_events()
